aquadoggo/src/network/dialer/behaviour.rs

```diff
@@ -0,0 +1,246 @@
+    /// no more redial attempts will occur.
+    fn schedule_dial(&mut self, peer_id: &PeerId) {
+        if let Some(status) = self.peers.get_mut(peer_id) {
+            if let Some(backoff_delay) = self.backoff.next(status.attempts) {
+                status.next_dial = Some(Instant::now() + backoff_delay);
+            } else {
+                debug!("Re-dial attempt limit reached: remove peer {peer_id:?}");
+                self.remove_peer(peer_id)
+            }
+        }
+    }
+
+    fn on_connection_established(&mut self, peer_id: PeerId) {
+        match self.peers.get_mut(&peer_id) {
+            Some(status) => {
+                status.connections += 1;
+            }
+            None => {
+                warn!("Connected established to unknown peer");
+            }
+        }
+    }
+
+    fn on_connection_closed(&mut self, peer_id: &PeerId) {
+        match self.peers.get_mut(peer_id) {
+            Some(status) => {
+                status.connections -= 1;
+
+                if status.connections == 0 {
+                    self.schedule_dial(peer_id)
+                }
+            }
+            None => {
+                warn!("Connected closed to unknown peer");
+            }
+        }
+    }
+
+    fn on_dial_failed(&mut self, peer_id: &PeerId) {
+        match self.peers.get_mut(peer_id) {
+            Some(status) => {
+                if status.connections == 0 {
+                    self.schedule_dial(peer_id)
+                }
+            }
+            None => warn!("Dial failed to unknown peer"),
+        }
+    }
+
+    fn remove_peer(&mut self, peer_id: &PeerId) {
+        if self.peers.remove(peer_id).is_none() {
+            // Don't warn here, peers can be removed twice if their registration on the swarm
+            // expired at the same time as their RETRY_LIMIT was reached.
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = DummyConnectionHandler;
+
+    type ToSwarm = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(DummyConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(DummyConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::ConnectionEstablished(ConnectionEstablished { peer_id, .. }) => {
+                self.on_connection_established(peer_id);
+            }
+            FromSwarm::ConnectionClosed(ConnectionClosed { peer_id, .. }) => {
+                self.on_connection_closed(&peer_id);
```

Comment:
This would try to redial even if the other peer closed the connection for good, right? Can we make this happen only on an error? Think for this we might need to call a behavior method from the swarm.

Reply:
Or is this dial condition taking care of that?

Reply:
Good point! I've accounted for this now with the re-write.